Test_snippet_set_present fails due to beta channel snippet returning 204 No Content

This test is failing due to the response being a 204, rather than a 302:

https://github.com/mozmeao/snippets-service/blob/6d7d7425187f9820c4e1faef5786f9a0dabf55f6/tests/test_snippets.py#L24

```
20:48:18 FAIL test_snippets.py::TestSnippets::()::test_snippet_set_present[/3/Firefox/default/default/default/en-US/beta/default/default/default/]
20:48:18 =================================== FAILURES ===================================
20:48:18  TestSnippets.test_snippet_set_present[/3/Firefox/default/default/default/en-US/beta/default/default/default/] 
20:48:18 [gw7] linux2 -- Python 2.7.5 /home/jenkins/slavehome/workspace/snippets.prod/tests/.tox/py27/bin/python2.7
20:48:18 
20:48:18 self = <test_snippets.TestSnippets instance at 0x2e9e908>
20:48:18 base_url = 'https://snippets.mozilla.com'
20:48:18 path = '/3/Firefox/default/default/default/en-US/beta/default/default/default/'
20:48:18 
20:48:18     @pytest.mark.parametrize(('path'), test_data)
20:48:18     def test_snippet_set_present(self, base_url, path):
20:48:18         full_url = base_url + path
20:48:18     
20:48:18         r = self._get_redirect(full_url)
20:48:18 >       assert requests.codes.ok == r.status_code, full_url
20:48:18 E       AssertionError: https://snippets.mozilla.com/3/Firefox/default/default/default/en-US/beta/default/default/default/
20:48:18 E       assert 200 == 204
20:48:18 E        +  where 200 = <lookup 'status_codes'>.ok
20:48:18 E        +    where <lookup 'status_codes'> = requests.codes
20:48:18 E        +  and   204 = <Response [204]>.status_code
20:48:18 
20:48:18 test_snippets.py:65: AssertionError
```

--- a/tests/test_snippets.py
+++ b/tests/test_snippets.py
@@ -18,10 +18,7 @@ REQUESTS_TIMEOUT = 20
 
 class TestSnippets:
 
-    test_data = [
-        ('/3/Firefox/default/default/default/en-US/release/default/default/default/'),
-        ('/3/Firefox/default/default/default/en-US/aurora/default/default/default/'),
-        ('/3/Firefox/default/default/default/en-US/beta/default/default/default/')]
+    URL_TEMPLATE = '{}/{}/Firefox/default/default/default/en-US/{}/default/default/default/'
 
     _user_agent_firefox = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.7; rv:10.0.1) Gecko/20100101 Firefox/10.0.1'
 
@@ -49,45 +46,59 @@ class TestSnippets:
                              allow_redirects=True, verify=False)
         except requests.exceptions.RequestException as e:
             raise AssertionError('Error connecting to {0} in {1}: {2}'.format(
-                url, path, e))
-        assert requests.codes.ok == r.status_code, \
-            'Bad URL {0} found in {1}'.format(url, path)
-        return True
+                url, e))
+        assert requests.codes.ok == r.status_code
 
     def _parse_response(self, content):
-        return BeautifulSoup(content)
+        return BeautifulSoup(content, 'html.parser')
 
-    @pytest.mark.parametrize(('path'), test_data)
-    def test_snippet_set_present(self, base_url, path):
-        full_url = base_url + path
+    @pytest.mark.parametrize(('version'), ['3', '5'])
+    @pytest.mark.parametrize(('channel'), ['aurora', 'beta', 'release'])
+    def test_response_codes(self, base_url, version, channel):
+        url = self.URL_TEMPLATE.format(base_url, version, channel)
+        r = self._get_redirect(url)
+        assert r.status_code in (requests.codes.ok, requests.codes.no_content)
 
-        r = self._get_redirect(full_url)
-        assert requests.codes.ok == r.status_code, full_url
-
-        soup = self._parse_response(r.content)
-        snippet_script = soup.find('script', type="text/javascript").text
-        snippet_json_string = re.search("JSON.parse\('(.+)'\)", snippet_script).groups()[0]
+    @pytest.mark.parametrize(('channel'), [
+        'aurora',
+        'beta',
+        pytest.param('release', marks=pytest.mark.xfail(
+            'mozilla' in pytest.config.getoption('base_url'),
+            reason='Activity Stream will be available after Firefox 57'))])
+    def test_activity_stream(self, base_url, channel):
+        url = self.URL_TEMPLATE.format(base_url, '5', channel)
+        soup = self._parse_response(self._get_redirect(url).content)
+        script = soup.find('script', type='application/javascript').text
+        snippet_json_string = re.search('JSON\.parse\("(.+)"\)', script).groups()[0]
         snippet_set = json.loads(snippet_json_string.replace('%u', r'\u').decode('unicode-escape'))
-
         assert isinstance(snippet_set, list), 'No snippet set found'
 
-    @pytest.mark.parametrize(('path'), test_data)
-    def test_all_links(self, base_url, path):
-        full_url = base_url + path
+    def test_legacy(self, base_url):
+        url = self.URL_TEMPLATE.format(base_url, '3', 'release')
+        soup = self._parse_response(self._get_redirect(url).content)
+        script = soup.find('script', type='text/javascript').text
+        snippet_json_string = re.search("JSON\.parse\('(.+)'\)", script).groups()[0]
+        snippet_set = json.loads(snippet_json_string.replace('%u', r'\u').decode('unicode-escape'))
+        assert isinstance(snippet_set, list), 'No snippet set found'
 
-        soup = self._parse_response(self._get_redirect(full_url).content)
+    @pytest.mark.parametrize(('version'), ['3', '5'])
+    @pytest.mark.parametrize(('channel'), ['aurora', 'beta', 'release'])
+    def test_all_links(self, base_url, version, channel):
+        url = self.URL_TEMPLATE.format(base_url, version, channel)
+
+        soup = self._parse_response(self._get_redirect(url).content)
         snippet_links = soup.select("a")
-
         for link in snippet_links:
-            self.assert_valid_url(link['href'], path)
+            self.assert_valid_url(link['href'])
 
-    @pytest.mark.parametrize(('path'), test_data)
-    def test_that_snippets_are_well_formed_xml(self, base_url, path):
-        full_url = base_url + path
+    @pytest.mark.parametrize(('channel'), ['aurora', 'beta', 'release'])
+    def test_that_snippets_are_well_formed_xml(self, base_url, channel):
+        url = self.URL_TEMPLATE.format(base_url, '3', channel)
 
-        r = self._get_redirect(full_url)
+        r = self._get_redirect(url)
         try:
+            print(r.content)
             parseString('<div>{}</div>'.format(r.content))
         except ExpatError as e:
             raise AssertionError('Snippets at {0} do not contain well formed '
-                                 'xml: {1}'.format(full_url, e))
+                                 'xml: {1}'.format(url, e))

--- a/tests/test_snippets.py
+++ b/tests/test_snippets.py
@@ -23,8 +23,7 @@ def _get_redirect(url, user_agent=_user_agent_firefox, locale='en-US'):
     headers = {'user-agent': user_agent,
                'accept-language': locale}
 
-    # verify=False ignores invalid certificate
-    return requests.get(url, headers=headers, verify=False, timeout=REQUESTS_TIMEOUT)
+    return requests.get(url, headers=headers, timeout=REQUESTS_TIMEOUT)
 
 
 def _parse_response(content):

--- a/tests/test_snippets.py
+++ b/tests/test_snippets.py
@@ -4,7 +4,6 @@
 
 import json
 import re
-from urlparse import urlparse
 from xml.dom.minidom import parseString
 from xml.parsers.expat import ExpatError
 
@@ -29,25 +28,6 @@ class TestSnippets:
 
         # verify=False ignores invalid certificate
         return requests.get(url, headers=headers, verify=False, timeout=REQUESTS_TIMEOUT)
-
-    def assert_valid_url(self, url, path, user_agent=_user_agent_firefox, locale='en-US'):
-        """Checks if a URL returns a 200 OK response."""
-
-        # Only check http and https links
-        if not urlparse(url).scheme.startswith('http'):
-            return True
-
-        headers = {'user-agent': user_agent,
-                   'accept-language': locale}
-
-        # HEAD doesn't return page body.
-        try:
-            r = requests.get(url, headers=headers, timeout=REQUESTS_TIMEOUT,
-                             allow_redirects=True, verify=False)
-        except requests.exceptions.RequestException as e:
-            raise AssertionError('Error connecting to {0} in {1}: {2}'.format(
-                url, e))
-        assert requests.codes.ok == r.status_code
 
     def _parse_response(self, content):
         return BeautifulSoup(content, 'html.parser')
@@ -80,16 +60,6 @@ class TestSnippets:
         snippet_json_string = re.search("JSON\.parse\('(.+)'\)", script).groups()[0]
         snippet_set = json.loads(snippet_json_string.replace('%u', r'\u').decode('unicode-escape'))
         assert isinstance(snippet_set, list), 'No snippet set found'
-
-    @pytest.mark.parametrize(('version'), ['3', '5'])
-    @pytest.mark.parametrize(('channel'), ['aurora', 'beta', 'release'])
-    def test_all_links(self, base_url, version, channel):
-        url = self.URL_TEMPLATE.format(base_url, version, channel)
-
-        soup = self._parse_response(self._get_redirect(url).content)
-        snippet_links = soup.select("a")
-        for link in snippet_links:
-            self.assert_valid_url(link['href'])
 
     @pytest.mark.parametrize(('channel'), ['aurora', 'beta', 'release'])
     def test_that_snippets_are_well_formed_xml(self, base_url, channel):

--- a/tests/test_snippets.py
+++ b/tests/test_snippets.py
@@ -13,62 +13,64 @@ import requests
 
 
 REQUESTS_TIMEOUT = 20
+URL_TEMPLATE = '{}/{}/Firefox/default/default/default/en-US/{}/default/default/default/'
+
+_user_agent_firefox = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.7; rv:10.0.1) Gecko/20100101 Firefox/10.0.1'
 
 
-class TestSnippets:
+def _get_redirect(url, user_agent=_user_agent_firefox, locale='en-US'):
 
-    URL_TEMPLATE = '{}/{}/Firefox/default/default/default/en-US/{}/default/default/default/'
+    headers = {'user-agent': user_agent,
+               'accept-language': locale}
 
-    _user_agent_firefox = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.7; rv:10.0.1) Gecko/20100101 Firefox/10.0.1'
+    # verify=False ignores invalid certificate
+    return requests.get(url, headers=headers, verify=False, timeout=REQUESTS_TIMEOUT)
 
-    def _get_redirect(self, url, user_agent=_user_agent_firefox, locale='en-US'):
 
-        headers = {'user-agent': user_agent,
-                   'accept-language': locale}
+def _parse_response(content):
+    return BeautifulSoup(content, 'html.parser')
 
-        # verify=False ignores invalid certificate
-        return requests.get(url, headers=headers, verify=False, timeout=REQUESTS_TIMEOUT)
 
-    def _parse_response(self, content):
-        return BeautifulSoup(content, 'html.parser')
+@pytest.mark.parametrize(('version'), ['3', '5'])
+@pytest.mark.parametrize(('channel'), ['aurora', 'beta', 'release'])
+def test_response_codes(base_url, version, channel):
+    url = URL_TEMPLATE.format(base_url, version, channel)
+    r = _get_redirect(url)
+    assert r.status_code in (requests.codes.ok, requests.codes.no_content)
 
-    @pytest.mark.parametrize(('version'), ['3', '5'])
-    @pytest.mark.parametrize(('channel'), ['aurora', 'beta', 'release'])
-    def test_response_codes(self, base_url, version, channel):
-        url = self.URL_TEMPLATE.format(base_url, version, channel)
-        r = self._get_redirect(url)
-        assert r.status_code in (requests.codes.ok, requests.codes.no_content)
 
-    @pytest.mark.parametrize(('channel'), [
-        'aurora',
-        'beta',
-        pytest.param('release', marks=pytest.mark.xfail(
-            'mozilla' in pytest.config.getoption('base_url'),
-            reason='Activity Stream will be available after Firefox 57'))])
-    def test_activity_stream(self, base_url, channel):
-        url = self.URL_TEMPLATE.format(base_url, '5', channel)
-        soup = self._parse_response(self._get_redirect(url).content)
-        script = soup.find('script', type='application/javascript').text
-        snippet_json_string = re.search('JSON\.parse\("(.+)"\)', script).groups()[0]
-        snippet_set = json.loads(snippet_json_string.replace('%u', r'\u').decode('unicode-escape'))
-        assert isinstance(snippet_set, list), 'No snippet set found'
+@pytest.mark.parametrize(('channel'), [
+    'aurora',
+    'beta',
+    pytest.param('release', marks=pytest.mark.xfail(
+        'mozilla' in pytest.config.getoption('base_url'),
+        reason='Activity Stream will be available after Firefox 57'))])
+def test_activity_stream(base_url, channel):
+    url = URL_TEMPLATE.format(base_url, '5', channel)
+    soup = _parse_response(_get_redirect(url).content)
+    script = soup.find('script', type='application/javascript').text
+    snippet_json_string = re.search('JSON\.parse\("(.+)"\)', script).groups()[0]
+    snippet_set = json.loads(snippet_json_string.replace('%u', r'\u').decode('unicode-escape'))
+    assert isinstance(snippet_set, list), 'No snippet set found'
 
-    def test_legacy(self, base_url):
-        url = self.URL_TEMPLATE.format(base_url, '3', 'release')
-        soup = self._parse_response(self._get_redirect(url).content)
-        script = soup.find('script', type='text/javascript').text
-        snippet_json_string = re.search("JSON\.parse\('(.+)'\)", script).groups()[0]
-        snippet_set = json.loads(snippet_json_string.replace('%u', r'\u').decode('unicode-escape'))
-        assert isinstance(snippet_set, list), 'No snippet set found'
 
-    @pytest.mark.parametrize(('channel'), ['aurora', 'beta', 'release'])
-    def test_that_snippets_are_well_formed_xml(self, base_url, channel):
-        url = self.URL_TEMPLATE.format(base_url, '3', channel)
+def test_legacy(base_url):
+    url = URL_TEMPLATE.format(base_url, '3', 'release')
+    soup = _parse_response(_get_redirect(url).content)
+    script = soup.find('script', type='text/javascript').text
+    snippet_json_string = re.search("JSON\.parse\('(.+)'\)", script).groups()[0]
+    snippet_set = json.loads(snippet_json_string.replace('%u', r'\u').decode('unicode-escape'))
+    assert isinstance(snippet_set, list), 'No snippet set found'
 
-        r = self._get_redirect(url)
-        try:
-            print(r.content)
-            parseString('<div>{}</div>'.format(r.content))
-        except ExpatError as e:
-            raise AssertionError('Snippets at {0} do not contain well formed '
-                                 'xml: {1}'.format(url, e))
+
+@pytest.mark.parametrize(('channel'), ['aurora', 'beta', 'release'])
+def test_that_snippets_are_well_formed_xml(base_url, channel):
+    url = URL_TEMPLATE.format(base_url, '3', channel)
+
+    r = _get_redirect(url)
+    try:
+        print(r.content)
+        parseString('<div>{}</div>'.format(r.content))
+    except ExpatError as e:
+        raise AssertionError('Snippets at {0} do not contain well formed '
+                             'xml: {1}'.format(url, e))

--- a/tests/test_snippets.py
+++ b/tests/test_snippets.py
@@ -31,7 +31,7 @@ def _parse_response(content):
     return BeautifulSoup(content, 'html.parser')
 
 
-@pytest.mark.parametrize(('version'), ['3', '5'])
+@pytest.mark.parametrize(('version'), ['3', '5'], ids=['legacy', 'activitystream'])
 @pytest.mark.parametrize(('channel'), ['aurora', 'beta', 'release'])
 def test_response_codes(base_url, version, channel):
     url = URL_TEMPLATE.format(base_url, version, channel)

--- a/tests/tox.ini
+++ b/tests/tox.ini
@@ -21,7 +21,7 @@ commands = flake8 {posargs:.}
 ignore = E501
 
 [pytest]
-addopts = --verbose -r=a
+addopts = --verbose -r=a --showlocals
 testpaths = ../tests
 xfail_strict = true
 base_url = https://snippets.allizom.org

--- a/tests/tox.ini
+++ b/tests/tox.ini
@@ -21,7 +21,7 @@ commands = flake8 {posargs:.}
 ignore = E501
 
 [pytest]
-addopts = --verbose -r=a --showlocals
+addopts = --verbose -r=a -n=auto --showlocals
 testpaths = ../tests
 xfail_strict = true
 base_url = https://snippets.allizom.org

